schdlr/workflow.py

```diff
@@ -0,0 +1,148 @@
+import random
+import string
+import networkx as nx
+
+from schdlr import etc
+from schdlr import task
+
+
+class DummyNode:
+
+    def __init__(self, name, done):
+        self.name = name
+        self.done = done
+        self.failed = False
+
+    @property
+    def state(self):
+        return task.DONE if self.done else task.PENDING
+
+    def __repr__(self):
+        return self.name
+
+
+def build_DAG(deps):
+    g = nx.DiGraph()
+
+    nodes = set()
+    dep_pairs = []
+
+    for v, deps_from in deps.items():
+        nodes.add(v)
+        for u in deps_from:
+            nodes.add(u)
+            dep_pairs.append((u, v))
+
+    for node in nodes:
+        g.add_node(node)
+
+    for v, u in dep_pairs:
+        g.add_edge(v, u)
+
+    start = DummyNode('START', True)
+    end = DummyNode('END', False)
+    g.add_node(start)
+    g.add_node(end)
+
+    for v in g.nodes:
+        if v in [start, end]:
+            continue
+        if not list(g.successors(v)):
+            g.add_edge(v, end)
+        if not list(g.predecessors(v)):
+            g.add_edge(start, v)
+
+    # verify that graph has valid workflow structure(DAG)
+    if not nx.has_path(g, start, end):
+        raise Exception("There is no path in generated graph from START to END")
+    if not nx.is_directed_acyclic_graph(g):
+        raise Exception("Generated graph is not a valid DAG")
+
+    return g, start, end
+
+
+NOT_STARTED = "NOT_STARTED"
+IN_PROGRESS = "IN_PROGRESS"
+FAILED = "FAILED"
+COMPLETED = "COMPLETED"
+
+
+class Workflow:
+
+    def __init__(self, deps, name=None):
+        """
+        :param deps: dependencies dict
+         task1: [task2,task3] -> task1 depends on task2 and task3
+        """
+        self.name = name if name else "".join(
+            [random.choice(string.ascii_letters) for _ in range(2)] +
+            [str(random.randint(0, 10)) for _ in range(2)]
+        )
+        self.deps = deps
+        self.g, self.start, self.end = build_DAG(self.deps)
+        self._state = NOT_STARTED
+
+        self.tasks_to_do = [self.start, ]
+        self.tasks_in_progress = []
+        self.tasks_done = []
+        self.logger = etc.get_logger("wflow.{name}".format(name=self.name))
+
+    def print(self):
+        import matplotlib.pyplot as plt
```

Comment:
Do you plan to rewrite such implementation in the future? From my point of view importing matplotlib is a little bit overcomplication for such method.

Reply:
Plan to remove it completely as soon as we will have webUI. Currently, It is the only way to draw a workflow.